test/integration/oauth_basicauth_test.go

```diff
@@ -288,7 +309,151 @@ func TestOAuthBasicAuthPassword(t *testing.T) {
 		t.Fatalf("Unexpected error: %v", err)
 	}
 	if user.Name != expectedUsername {
-		t.Fatalf("Expected username as the user, got %v", user)
+		t.Fatalf("Expected %v as the user, got %v", expectedUsername, user)
+	}
+
+}
+
+func launchBasicAuthRedirectingServer(t *testing.T, certNames map[string]string, redirectCode int) (string, error) {
+	realLoginPage := "/real_login_page"
+
+	// Start remote server
+	remoteAddr, err := testserver.FindAvailableBindAddress(9443, 9999)
+	if err != nil {
+		return "", err
+	}
+
+	// Build client cert pool
+	clientCAs, err := util.CertPoolFromFile(certNames[basicAuthRemoteCACert])
+	if err != nil {
+		return "", err
+	}
+	// Build remote handler
+	redirectHandler := http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if req.TLS == nil {
+			w.WriteHeader(http.StatusUnauthorized)
+			t.Fatalf("Expected TLS")
+		}
+		if len(req.TLS.VerifiedChains) != 1 {
+			w.WriteHeader(http.StatusUnauthorized)
+			t.Fatalf("Expected peer cert verified by server")
+		}
+		if req.URL.Path == realLoginPage {
+			w.Header().Set("Content-Type", "application/json")
+			w.Write([]byte(fmt.Sprintf(`{"sub":"%s"}`, expectedUsername)))
+			return
+		}
+
+		//Return a redirect with the status code provided
+		http.Redirect(w, req, realLoginPage, redirectCode)
+		return
+	})
+
+	remoteServer := &http.Server{
+		Addr:           remoteAddr,
+		Handler:        redirectHandler,
+		ReadTimeout:    10 * time.Second,
+		WriteTimeout:   10 * time.Second,
+		MaxHeaderBytes: 1 << 20,
+		TLSConfig: crypto.SecureTLSConfig(&tls.Config{
+			// RequireAndVerifyClientCert lets us limit requests to ones with a valid client certificate
+			ClientAuth: tls.RequireAndVerifyClientCert,
+			ClientCAs:  clientCAs,
+		}),
+	}
+
+	go func() {
+		if err := remoteServer.ListenAndServeTLS(certNames[basicAuthRemoteServerCert], certNames[basicAuthRemoteServerKey]); err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+	}()
+
+	return remoteAddr, nil
+}
+
+func TestOAuthBasicAuthPasswordRedirect301(t *testing.T) {
+	runRedirectTests(t, 301)
+}
+func TestOAuthBasicAuthPasswordRedirect302(t *testing.T) {
+	runRedirectTests(t, 302)
+}
+func TestOAuthBasicAuthPasswordRedirect303(t *testing.T) {
+	runRedirectTests(t, 303)
+}
+func TestOAuthBasicAuthPasswordRedirect307(t *testing.T) {
+	runRedirectTests(t, 307)
```

Comment:
So are you suggesting I just pick one of the codes and test that, or modify the test so I only start and stop a single instance of OpenShift?